src/Processors/QueryPlan/Optimizations/calculateHashTableCacheKeys.cpp

```diff
@@ -0,0 +1,199 @@
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+
+#include <Core/Joins.h>
+#include <IO/WriteHelpers.h>
+#include <Interpreters/SetSerialization.h>
+#include <Interpreters/TableJoin.h>
+#include <Processors/QueryPlan/ITransformingStep.h>
+#include <Processors/QueryPlan/JoinStepLogical.h>
+#include <Processors/QueryPlan/ReadFromRemote.h>
+#include <Processors/QueryPlan/Serialization.h>
+#include <Processors/QueryPlan/SourceStepWithFilter.h>
+#include <Storages/IStorage.h>
+#include <Common/SipHash.h>
+
+using namespace DB;
+
+namespace
+{
+
+UInt64 calculateHashFromStep(const ReadFromParallelRemoteReplicasStep & source)
+{
+    SipHash hash;
+    hash.update(source.getSerializationName());
+    hash.update(source.getStorageID().getFullTableName());
+    return hash.get64();
+}
+
+UInt64 calculateHashFromStep(const SourceStepWithFilter & read)
+{
+    WriteBufferFromOwnString wbuf;
+    SerializedSetsRegistry registry;
+    IQueryPlanStep::Serialization ctx{.out = wbuf, .registry = registry};
+
+    writeStringBinary(read.getSerializationName(), wbuf);
+    if (const auto & snapshot = read.getStorageSnapshot())
+        writeStringBinary(snapshot->storage.getStorageID().getFullTableName(), wbuf);
+    if (const auto & dag = read.getPrewhereInfo())
+        dag->prewhere_actions.serialize(ctx.out, ctx.registry);
+
+    SipHash hash;
+    hash.update(wbuf.str());
+    return hash.get64();
+}
+
+UInt64 calculateHashFromStep(const ITransformingStep & transform)
+{
+    if (!transform.getTransformTraits().preserves_number_of_rows)
```

Comment:
This condition is not very clear to me.
I guess we ignore steps that do not filter anything, but the comment would be helpful.

Also, if we return 0 and update the hash with it, the number of steps would matter. E.g. hash for `Expression -> Expresion` would be different than hash from just 1 Expression, etc.

Reply:
> the number of steps would matter

That's a great point, thx